### PR TITLE
`setequal()` improvements

### LINKED ISSUE #6114: setdiff vs. setequal inconsitency when columns have different types
It seems like `setequal` compares column types while `setdiff` doesn't.

```r
df1 <- data.frame(x=1)
df2 <- df1

# df1$x and df2$x are both of type `num`
setdiff(df1, df2) # --> will output 0-rowed data.frame
setequal(df1, df2) # --> will output TRUE

df2$x %<>% as.integer 

# df1$x is of type `num` while df2$x is of type `int`
setdiff(df1, df2) # --> will output 0-rowed data.frame
setequal(df1, df2) # --> will output FALSE
```

I can see how this could actually be the intended behavior. 
However, in that case a hint in the docs would be great.

### PR DESCRIPTION
**Pull request by hadley**
I don't love that there's so much variation in where the casting happens in these functions, but this seems like the simplest fix. `convert = TRUE` here eventually calls `vec_ptype2(x_i, y_i)` so it's still using vctrs rules.

Fixes #6114.

### CODE REVIEW

**Comment by hadley** [comment]
I also noticed #6057, and fixing that leads to a cleaner implementation, I think.

**Comment by DavisVaughan** [comment]
Closes #6057

File: R/sets.r [hunk 1]
```diff
@@ -67,44 +67,51 @@ generics::setequal
...
+  out <- vec_unique(vec_rbind(x, y))
   reconstruct_set(out, x)
 }
 
 #' @export
 union_all.data.frame <- function(x, y, ...) {
   check_dots_empty()
-  out <- bind_rows(x, y)
+
+  out <- vec_rbind(x, y)
```
Comment (DavisVaughan) [inline R/sets.r:89]: Should this call `check_compatible(x, y)` like `union()`? i.e. to me `union() == vec_unique(union_all())`?

File: R/sets.r [hunk 2]
```diff
@@ -67,44 +67,51 @@ generics::setequal
...
   check_dots_empty()
   check_compatible(x, y)
-  cast <- vec_cast_common(x, y)
-  new_x <- cast[[1L]]
-  new_y <- cast[[2L]]
-  out <- vec_unique(vec_slice(new_x, !vec_in(new_x, new_y)))
+
+  cast <- vec_cast_common(x = x, y = y)
+  out <- vec_unique(vec_slice(cast$x, !vec_in(cast$x, cast$y)))
   reconstruct_set(out, x)
```
Comment (DavisVaughan) [inline R/sets.r:100]: Is this `dplyr_reconstruct()`? It certainly looks like it if you compare to `dplyr_reconstruct.grouped_df`

File: R/sets.r [hunk 3]
```diff
@@ -67,44 +67,51 @@ generics::setequal
...
 }
 
 #' @export
 setequal.data.frame <- function(x, y, ...) {
   check_dots_empty()
-  isTRUE(equal_data_frame(x, y))
+  if (!is.data.frame(y)) {
+    abort("`y` must be a data frame. ", call = error_call)
```
Comment (DavisVaughan) [inline R/sets.r:107]:
```suggestion
    abort("`y` must be a data frame.")
```

No error call there, plus an extra space. Maybe add a test for this?

**Comment by hadley** [comment]
I don't fully understand the logic behind "linked issues". You'd think it'd be the same as the issues that get closed when the PR is merged but it's not (I already had `Closes #6057` in the commit message).

File: tests/testthat/test-sets.R [hunk 4]
```diff
@@ -90,6 +93,32 @@ test_that("set equality", {
...
+
+test_that("setequal equality ignores duplicated rows", {
+  df1 <- tibble(x = 1)
+  df2 <- df1[c(1, 1, 1), ]
+
+  expect_true(setequal(df1, df2))
+  expect_true(setequal(df2, df1))
+})
+
+test_that("setequal uses coercion rules", {
```
Comment (DavisVaughan) [inline tests/testthat/test-sets.R:112]:
```suggestion
test_that("setequal uses coercion rules (#6114)", {
```

File: tests/testthat/test-sets.R [hunk 5]
```diff
@@ -90,6 +93,32 @@ test_that("set equality", {
...
 
+test_that("setequal ignores column and row order", {
+  df1 <- tibble(x = 1:2, y = 3:4)
+  df2 <- df1[2:1, 2:1]
+
+  expect_true(setequal(df1, df2))
+  expect_true(setequal(df1, df2))
+})
+
+test_that("setequal equality ignores duplicated rows", {
```
Comment (DavisVaughan) [inline tests/testthat/test-sets.R:104]: There is an existing (but odd) test above called `"set operations remove duplicates"` if you wanted to add to that

Comment (DavisVaughan) [inline tests/testthat/test-sets.R:104]:
```suggestion
test_that("setequal equality ignores duplicated rows (#6057)", {
```

File: R/sets.r [hunk 6]
```diff
@@ -67,44 +67,51 @@ generics::setequal
...
+
   cast <- vec_cast_common(x = x, y = y)
-  new_x <- cast[[1L]]
-  new_y <- cast[[2L]]
-  out <- vec_unique(vec_slice(new_x, vec_in(new_x, new_y)))
+  out <- vec_unique(vec_slice(cast$x, vec_in(cast$x, cast$y)))
   reconstruct_set(out, x)
 }
 
 #' @export
```
Comment (DavisVaughan) [inline R/sets.r:76]: Random note that the help docs for these could benefit from these methods getting a `#' @rdname setops` tag

If you do `?union_all` the help page only has Usage for `union_all()` even though the help page is for all the methods

**Review by DavisVaughan — APPROVED** [review]

File: R/sets.r (hunk 6 above)
Comment (hadley) [reply R/sets.r:76]: See my other docs PR

File: R/sets.r (hunk 1 above)
Comment (hadley) [reply R/sets.r:89]: Yeah, definitely

File: R/sets.r (hunk 2 above)
Comment (hadley) [reply R/sets.r:100]: Yeah, good catch. Must've be written before `dplyr_reconstruct` existed

File: R/sets.r (hunk 3 above)
Comment (hadley) [reply R/sets.r:107]: Oops yeah.

File: tests/testthat/test-sets.R (hunk 5 above)
Comment (hadley) [reply tests/testthat/test-sets.R:104]: I decided to put all the setequal tests together since it's a bit different. I'll do a little more test cleanup.

**Pull request merged**